shared/modules/metametrics.js

```diff
@@ -0,0 +1,196 @@
+ * @property {string}  event - event name to track
+ * @property {string}  category - category to associate event to
+ * @property {boolean} [isOptIn] - happened during opt in/out workflow
+ * @property {object}  [properties] - object of custom values to track, snake_case
+ * @property {number}  [revenue] - amount of currency that event creates in revenue for MetaMask
+ * @property {string}  [currency] - ISO 4127 format currency for events with revenue, defaults to US dollars
+ * @property {number}  [value] - Abstract "value" that this event has for MetaMask.
+ * @property {boolean} [excludeMetaMetricsId] - whether to exclude the user's metametrics id for anonymity
+ * @property {MetaMetricsEventContext} [additionalContext] - additional context to attach to event
+ */
+
+/**
+ * Returns a function for tracking Segment events.
+ *
+ * @param {string} metamaskVersion - The current version of the MetaMask extension.
+ * @param {() => MetaMetricsRequiredState} getDynamicState - A function returning required fields
+ * @param {boolean} defaultMetaMetricsExclusion - Used to set the default state of exludeMetaMetricsId
+ * @returns {(payload: MetaMetricsEventPayload) => Promise<never>} - function to track an event
+ */
+export function getTrackMetaMetricsEvent (
+  metamaskVersion,
+  getDynamicState,
+  defaultMetaMetricsExclusion = false,
+) {
+  const version = process.env.METAMASK_ENVIRONMENT === 'production'
+    ? metamaskVersion
+    : `${metamaskVersion}-${process.env.METAMASK_ENVIRONMENT}`
+
+  return function trackSegmentEvent ({
+    event,
+    category,
+    isOptIn,
+    properties = {},
+    revenue,
+    currency,
+    value,
+    excludeMetaMetricsId: excludeId,
+    additionalContext = {},
+  }) {
+    if (!event || !category) {
+      throw new Error('Must specify event and category.')
+    }
+    const {
+      participateInMetaMetrics,
+      context: providedContext,
+      metaMetricsId,
+      environmentType,
+      chainId,
+      network,
+      metaMetricsSendCount,
+    } = getDynamicState()
+
+    let excludeMetaMetricsId = excludeId ?? defaultMetaMetricsExclusion
+
+    // This is carried over from the old implementation, and will likely need
+    // to be updated to work with the new tracking plan. I think we should use
+    // a config setting for this instead of trying to match the event name
+    const isSendFlow = Boolean(event.match(/^send|^confirm/u))
+    if (isSendFlow && metaMetricsSendCount && !sendCountIsTrackable(metaMetricsSendCount + 1)) {
+      excludeMetaMetricsId = true
+    }
+
+    if (!participateInMetaMetrics && !isOptIn) {
+      return Promise.resolve()
+    }
+
+    const context = merge({
+      app: {
+        name: 'MetaMask Extension',
+        version,
+      },
+      userAgent: window.navigator.userAgent,
+    }, providedContext, additionalContext)
+
+    const trackOptions = {
+      event,
+      properties: {
+        ...omit(properties, ['revenue', 'currency', 'value']),
```

Comment:
Do we need the `omit` call here? The keys omitted from `properties` will be overwritten anyway since they are explicitly specified below.

Reply:
the keys below come from a different source. If someone tries to use these properties, which have special meaning in segment, I want to make them think about the usage. Ideally, when typescript comes around they'll get an immediate error when passing these keys to properties object, but for now, their payload will *not* include the values. When they investigate they will see that these three are special cases and should be elevated outside of the properties object. They might also realize that 'value' isn't an appropriate place to put the data they are tracking.

Reply:
> When they investigate they will see that these three are special cases

I think we could achieve the same effect with a comment!

Reply:
I added a comment above the omit. I feel strongly that we should avoid treating these as normal properties, but adding additional comments about why they are being omitted makes sense to me.